packages/tenant-process/src/services/validators.ts

```diff
@@ -46,6 +50,86 @@ export function assertVerifiedAttributeExistsInTenant(
   }
 }
 
+export async function assertVerifiedAttributeOperationAllowed({
+  producerId,
+  consumerId,
+  attributeId,
+  agreementStates,
+  readModelService,
+  error,
+}: {
+  producerId: string;
+  consumerId: string;
+  attributeId: string;
+  agreementStates: AgreementState[];
+  readModelService: ReadModelService;
+  error: Error;
+}): Promise<void> {
+  // Get agreements
+  const agreements = await readModelService.getAllAgreements({
+    producerId,
+    consumerId,
+    agreementStates,
+  });
+
+  // Extract descriptor IDs
+  const descriptorIds = agreements.map((agreement) => agreement.descriptorId);
+
+  // Get eServices concurrently
+  const eServices = (
+    await Promise.all(
+      agreements.map((agreement) =>
+        readModelService.getEServiceById(agreement.eserviceId)
+      )
+    )
+  ).filter((eService): eService is EService => eService !== undefined);
+
+  // Find verified attribute IDs
+  const attributeIds = eServices
+    .flatMap((eService) =>
+      eService.descriptors.filter((descriptor) =>
+        descriptorIds.includes(descriptor.id)
+      )
+    )
+    .flatMap((descriptor) =>
+      descriptor.attributes.verified.flatMap((attribute) =>
+        attribute.map((a) => a.id)
+      )
+    )
+    .reduce((acc, id) => acc.add(id), new Set<string>()); // Use Set for uniqueness
+
+  // Check if attribute is allowed
+  if (!attributeIds.has(attributeId)) {
+    throw error;
+  }
+}
+
+export async function assertAttributeVerificationAllowed({
+  producerId,
+  consumerId,
+  attributeId,
+  readModelService,
+}: {
+  producerId: TenantId;
+  consumerId: TenantId;
+  attributeId: AttributeId;
+  readModelService: ReadModelService;
+}): Promise<void> {
+  const allowedStatuses = [
+    agreementState.pending,
+    agreementState.active,
+    agreementState.suspended,
+  ];
+  await assertVerifiedAttributeOperationAllowed({
+    producerId,
+    consumerId,
+    attributeId,
+    agreementStates: allowedStatuses,
+    readModelService,
+    error: attributeVerificationNotAllowed(consumerId, attributeId),
+  });
+}
```

Comment:
Please double-check if the two levels of nested assertions are really required or can be simplified.